File: leetcode/algorithms/stack/minimum-add-to-make-parentheses-valid.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Minimum Add to Make Parentheses Valid"
*/


class Solution {
    public int minAddToMakeValid(String S) {
        Stack<Character> parenStack = new Stack<Character>();       
        for(int i=0;i<S.length();++i){
            Integer topCharValue = parenStack.empty() ? null : (int)parenStack.peek();
            if(topCharValue!=null && ((int)(S.charAt(i) - topCharValue)>=1)){
                parenStack.pop();
            } else {
                parenStack.push(S.charAt(i));
            }
        }
        return parenStack.size();
    }
}

File: leetcode/algorithms/stack/minimum-add-to-make-parentheses-valid.js
/*
    leetcode
    category: "Algorithms"
    problem title: "Minimum Add to Make Parentheses Valid"
*/


/**
 * @param {string} S
 * @return {number}
 */

const _PRIVATE_STACK_MAP = new WeakMap();

const Stack = function(){
    const data = {
        stack: []
    }
    _PRIVATE_STACK_MAP.set(this,data);
}

Stack.prototype.push = function(val){
    const data = _PRIVATE_STACK_MAP.get(this);
    data.stack.push(val);
}

Stack.prototype.pop = function(){
    const data = _PRIVATE_STACK_MAP.get(this);
    return data.stack.length>0 ? data.stack.pop() : null;
}

Stack.prototype.peek = function(){
    const data = _PRIVATE_STACK_MAP.get(this);
    return data.stack.length>0 ? data.stack[data.stack.length-1] : null;
}

Stack.prototype.getSize = function(){
    const data = _PRIVATE_STACK_MAP.get(this);
    return data.stack.length;
}

var minAddToMakeValid = function(S) {
    const parenStack = new Stack();
    for(let i=0;i<S.length;++i){
        if(parenStack.peek()!=null && S[i].charCodeAt(0)-parenStack.peek().charCodeAt(0)>=1){
            parenStack.pop();
        } else {
            parenStack.push(S[i]);
        }
    }
    return parenStack.getSize();
};